### PR TITLE
Added option for http response headers

### LINKED ISSUE #174: Provide security options for response header
These security options cannot be set for the response header:
* x-content-type-options
* x-xss-protection
* x-frame-options
* content-security-policy
* strict-transport-security

### LINKED ISSUE #334: option to set custom response headers, example Service-Worker-Allowed
option to set custom response headers, for instance  `Service-Worker-Allowed`. possible add more then one custom headers to response.

### LINKED ISSUE #644: Setting CSP
Is there a way to set CSP in a similar way to setting CORS, or more generically, is there a way to manually specify a batch of headers to send with each response? Having to actually write a complete Node based server solution just to test that a page works with CSP in place is really silly if `http-server` would allow that kind of testing with a one line, e.g.

```
> http-server -p 8000 --cors --csp="default-src 'self'; img-src *; media-src media1.com media2.com; script-src userscripts.example.com"
```

### PR DESCRIPTION
**Pull request by hedefalk**
Use case was: https://github.com/indexzero/http-server/issues/281

Feel free to close if doesn't seem relevant…

Fixes #174, fixes #334, fixes #644

### CODE REVIEW

**Comment by ssimono** [comment]
Hello, I just comment to say it seems totally relevant to me, I would really appreciate if it got merged.

**Comment by nojvek** [comment]
This is super useful. Someone please get it merged and do a publish. I could use this right now

**Comment by howardroark** [comment]
Yeah I think I could use this over what I submitted for #288 :)

**Comment by JamieMason** [comment]
Subscribed, I think this would allow me to set `AMP-Access-Control-Allow-Source-Origin` to develop https://www.ampproject.org projects.

**Comment by BigBlueHat** [comment]
👍 This would also useful for setting/testing Content Security Policy headings.

**Comment by jaylinski** [comment]
@indexzero any plan on merging this?

**Comment by coderkd10** [comment]
@indexzero gentle reminder. it will be super helpful if you could merge this PR. Also let me know if any changes are required. I'll be happy to help 😄

**Comment by Tofandel** [comment]
Only 5 years have passed...

**Comment by oncletom** [comment]
`minimist` allows the parsing of 1. Objects and 2. Multiple options of the same name.

Thus, this would work with minimal code addition:

```bash
http-server --headers.AMP-Access-Control-Allow-Source-Origin 'amp@gmail.dev' --headers.Cookie 'test=1'
```

The `argv.headers` would be equal to:

```js
{
  'AMP-Access-Control-Allow-Source-Origin': 'amp@gmail.dev',
  'Cookie': 'test=1'
}
```

Techically, that's the only thing to pass to the `options` object, as the `httpServer` already handles the (undocumented) `headers` option.

**Comment by thornjad** [comment]
@oncletom you're totally right about `minimist` supporting object options like that, however, I do like the PR's implementation, which makes an array, a little better because it passes the headers as a single string, which seems more natural to me.

**Review by thornjad — CHANGES_REQUESTED** [review]
Love the simplicity! Here's what we still need before merging:

- [ ] Add documentation to doc/http-server.1
- [ ] Add documentation to README.md
- [ ] Add tests (at least one for a single extra header and one for multiple extra headers)

File: lib/http-server.js
Comment (zbynek) [inline lib/http-server.js:81]:
```suggestion
      var split = header.split(':', 2);
```
might be a more obvious way to do this.

Comment (thornjad) [reply lib/http-server.js:81]: colons are allowed in the value part of the header (e.g. `Content-Type: Foo: 2: Tokyo Drift` is valid), so we'd need to concat everything after the first match again anyway. I would have though of doing it this way first too, but the regex capture group actually ends up being more terse, which is neat

Comment (zbynek) [reply lib/http-server.js:81]: @thornjad that's why I suggested passing `2` as 2nd argument to the `split(separator, limit)` syntax.

Comment (thornjad) [reply lib/http-server.js:81]:
```js
'Content-Type: Foo: 2: Tokyo Drift'.split(':', 2)
> Array [ "Content-Type", " Foo" ]
```

The limit just sets the max length of the resulting array

Comment (zbynek) [reply lib/http-server.js:81]: :facepalm: you're right, sorry.

Comment (thornjad) [reply lib/http-server.js:81]: No problem! The way you were thinking is the way it works in other languages like Python and Java, JS is the weird one out, as it so often is

**Comment by oncletom** [comment]
I’m game on providing the requested changes if @hedefalk somewhat cannot manage to do so.

**Comment by thornjad** [comment]
@oncletom would you like to make the requested changes? Otherwise I can take this over

**Comment by hlolli** [comment]
such a simple addition, surprised it's missing, wanted to add Cross-Origin-Resource-Policy to https://developer.chrome.com/blog/enabling-shared-array-buffer/#cross-origin-isolation

**Review by dzierzanowski — APPROVED** [review]
Looks good to me!

**Comment by rgeraldporter** [comment]
Would love to see this added.

**Comment by cwtuan** [comment]
is custom response headers supported?

**Comment by JulianIsrael** [comment]
would this allow custom headers for files matching a regex?

`-H *.js.gz="Content-Type=text/javascript" `

**Comment by Pomax** [comment]
@thornjad would it be possible to get this landed still?

**Comment by BelgianNoise** [comment]
Is this the sign for us to choose another package ?

**Comment by BigBlueHat** [comment]
@josenicomaia @BelgianNoise the best way to help any OSS project is to contribute. In this case, testing this PR, confirming that it works, explaining what you tested, etc. would go along way to helping the core contributors know if this patch is what's needed or if more work remains.

@cwtuan @JulianIsrael if you're curious about specific capabilities, please also take the time to test this branch, find out for yourselves, and provide confirmation here that it does support those things.

Community projects take a community.

**Comment by zbynek** [comment]
@BigBlueHat thanks for confirming that you're still interested in this project. Will you be ale to merge an release https://github.com/http-party/http-server/pull/811 ? (sorry for cross-posting, but that issue seems to be abandoned by maintainers)

**Comment by Tofandel** [comment]
To answer @cwtuan that's literally the point of the PR so I don't get the question

To answer @JulianIsrael nope it wouldn't there is nothing in this PR that would do that, but feel free to make a superseeding PR with the feature

As it was established, this PR is still missing a test case and documentation


Nice to know the project is still alive though

**Comment by alexanderankin** [comment]
> helping the core contributors know if this patch is what's needed or if more work remains

a PR being submitted usually implies that the submitter just wants to merge the changes supplied (unless explicitly noted otherwise). Open Source maintainers aren't supposed to be HR managers who sign off on things unknowingly?

as far as the requested changes, here's what I'd add to `docs/http-server.1`

```groff
.TP
.BI \-H ", " \-\-header " " \fIHEADER\fR
Add extra header to all responses, eg. "X-Frame-Options: DENY".
```

(right after cors):

```diff
--- man.1	2022-08-28 17:38:33.244404883 -0400
+++ man-before	2022-08-28 17:36:54.973302056 -0400
@@ -63,10 +63,6 @@
 Optionally provide CORS headers list separated by commas.
 
 .TP
-.BI \-\-H ", " \-\-header " " \fIHEADER\fR
-Add extra header to all responses, eg. "X-Frame-Options: DENY".
-
-.TP
 .BI \-o " " [\fIPATH\fR]
 Open default browser window after starting the server.
 Optionally provide a URL path to open the browser window to.
```

and to `README.md`:

```diff
--- README.md	2022-08-28 17:40:21.619418290 -0400
+++ README.md-after	2022-08-28 17:41:32.794771710 -0400
@@ -53,6 +53,7 @@
 |`-e` or `--ext`  |Default file extension if none supplied |`html` | 
 |`-s` or `--silent` |Suppress log messages from output  | |
 |`--cors` |Enable CORS via the `Access-Control-Allow-Origin` header  | |
+|`-H` or `--header` |Add extra header to all responses, eg. `X-Frame-Options: DENY`| |
 |`-o [path]` |Open browser window after starting the server. Optionally provide a URL path to open. e.g.: -o /other/dir/ | |
 |`-c` |Set cache time (in seconds) for cache-control max-age header, e.g. `-c10` for 10 seconds. To disable caching, use `-c-1`.|`3600` |
 |`-U` or `--utc` |Use UTC time format in log messages.| |
```

**Comment by alexanderankin** [comment]
here are tests:

```diff
diff --git a/test/http-server-test.js b/test/http-server-test.js
index d4555b9..d3c7f13 100644
--- a/test/http-server-test.js
+++ b/test/http-server-test.js
@@ -154,5 +154,48 @@ vows.describe('http-server').addBatch({
         assert.ok(res.headers['access-control-allow-headers'].split(/\s*,\s*/g).indexOf('X-Test') >= 0, 204);
       }
     }
+  },
+  'When extra header are provided': {
+    topic: function () {
+      var server = httpServer.createServer({
+        root: root,
+        extraHeaders: ['X-Header:extra']
+      });
+      server.listen(8083);
+      this.callback(null, server);
+    },
+    'and request is issued to server with extra header': {
+      topic: function () {
+        request({
+          method: 'GET',
+          uri: 'http://127.0.0.1:8083/'
+        }, this.callback);
+      },
+      'response should contain extra header': function (err, res) {
+        assert.equal(res.headers['x-header'], 'extra');
+      }
+    }
+  },
+  'When multiple extra headers are provided': {
+    topic: function () {
+      var server = httpServer.createServer({
+        root: root,
+        extraHeaders: ['X-Header-One:extra-one', 'X-Header-Two:extra-two']
+      });
+      server.listen(8084);
+      this.callback(null, server);
+    },
+    'and request is issued to server with extra headers': {
+      topic: function () {
+        request({
+          method: 'GET',
+          uri: 'http://127.0.0.1:8084/'
+        }, this.callback);
+      },
+      'response should contain extra headers': function (err, res) {
+        assert.equal(res.headers['x-header-one'], 'extra-one');
+        assert.equal(res.headers['x-header-two'], 'extra-two');
+      }
+    }
   }
 }).export(module);
```

**Comment by Pomax** [comment]
> a PR being submitted usually implies that the submitter just wants to merge the changes supplied (unless explicitly noted otherwise). Open Source maintainers aren't supposed to be HR managers who sign off on things unknowingly?

This PR addresses an open issue, it's not a drive-by PR. The discussion about what the PR should do goes in that issue, and the discussion in the PR itself should be able whether the code addresses the linked issue, and whether it requires changes before it can be merged or not. A PR being submitted does not imply someone "just wants to merge the changes", it implies they took on the task of fixing an issue and they are now submitting the work they believe does that, requiring review and sign-off from the people who actually maintain the project, as it's their responsibility to accept contributions or not.

Also, in terms of comments with diffs: if these are changes to files already in the PR, just use the PR commenting system instead (i.e. find the file and line(s) in the PR you flagged as needing more work, click the comment button next to th(os)e line number(s) and leave a comment with "code suggestion" that the PR owner can then automatically accept into the PR) or if it's a completely new change to a file not already in the PR, hit up their branch, click "edit" on that file, let github do its forking magic if necessary, and then file your own PR against their PR. This can all be done without ever leaving github.

**Comment by jonathanding** [comment]
Is there any update for this feature? This seems to be very useful and surprisingly to see it is missing and takes so long to get merged.

**Comment by 64jcl** [comment]
Was surprised there was no way to add custom headers to the response of http-server. I need it in order to add "Cross-Origin-Opener-Policy: same-origin" and "Cross-Origin-Embedder-Policy: require-corp" in order to get SharedArrayBuffer to work.

**Comment by iliakan** [comment]
> Was surprised there was no way to add custom headers to the response of http-server. I need it in order to add "Cross-Origin-Opener-Policy: same-origin" and "Cross-Origin-Embedder-Policy: require-corp" in order to get SharedArrayBuffer to work.

Exactly the same thing! There's no support yet?

**Comment by ddresch** [comment]
@iliakan if you are still struggling. Just merge the changes of the two files yourself. 

ATTENTION: not update safe – but anyway if just want to go on :)

Here is the process I used for PNPM:

1. find where the bin file is located on your system `which http-server`
2. to get your `basepath` edit bin file `nano [your-path]`
3. add this echo to print out your `basepath`
```bash
#!/bin/sh
basedir=$(dirname "$(echo "$0" | sed -e 's,\\,/,g')")
echo $basedir
```
4. open the directory with your editor
```
code [your-basepath]../pnpm-global/5/node_modules/http-server/bin/http-server
```
5. copy the changes from [files](https://github.com/http-party/http-server/pull/282/files)

**Comment by Tofandel** [comment]
6. Install https://www.npmjs.com/package/patch-package and run it on the http-server

**Comment by alshdavid** [comment]
For those still waiting on custom headers, I have been rewriting `http-server` in Rust with support for custom headers: [http-server-rs](https://github.com/alshdavid/http-server-rs)

**Comment by Pomax** [comment]
alternatively, most of my projects just use a copy-past `server.js` file based on express that just sets the entire project dir as a static dir and redirects `/` to index.html so that "things just work(tm)".

```lang-js
import express from "express";

const PORT = process.env.PORT ?? 8000;
process.env.PORT = PORT;

const HOSTNAME = process.env.HOSTNAME ?? `localhost`;
process.env.HOSTNAME = HOSTNAME;

const npm = process.platform === `win32` ? `npm.cmd` : `npm`;

// Set up the core server
const app = express();
app.use((req, res, next) => {
  res.setHeader("Surrogate-Control", "no-store");
  res.setHeader(
    "Cache-Control",
    "no-store, no-cache, must-revalidate, proxy-revalidate"
  );
  res.setHeader("Expires", "0");
  next();
});
app.set("etag", false);
app.use((req, res, next) => {
  if (!process.argv.includes(`--test`)) {
    console.log(`[${new Date().toISOString()}] ${req.url}`);
  }
  next();
});

// static routes
app.get(`/`, (req, res) => res.redirect(`/index.html`));
app.use(`/`, express.static(`.`));
app.use((req, res) => res.status(404).send(`${req.url} not found`));

// Run the server
app.listen(PORT, () => {
  const msg = `=   Server running on http://${HOSTNAME}:${PORT}   =`;
  const line = `=`.repeat(msg.length);
  const mid = `=${` `.repeat(msg.length - 2)}=`;
  console.log([``, line, mid, msg, mid, line, ``].join(`\n`));
});
```

Done, `node server` anywhere you would have used `http-server`.

**Comment by zbynek** [comment]
Also https://www.npmjs.com/package/serve is a good replacement for this package.

**Comment by indexzero** [comment]
Closing in favor of #887 / #885

**Comment by Pomax** [comment]
Probably better to leave this open until one of those _actually_ lands, given how little has happened in this project. I'll be happy to see either of those land, but so far the track record has been "that's not happening" =)

**Comment by KernelDeimos** [comment]
Newer PRs are easier to validate and merge, so this seems like reasonable decision to me. This thread you replied on is also a PR; issue #884 tracks this. I recently (today) became a maintainer which hopefully means I'll do a fantastic job at it and bring life back into the PRs - time will tell!

**Comment by Pomax** [comment]
I see! In that case, good luck, good to hear someone's back in the maintainer seat.

**Comment by KernelDeimos** [comment]
Anyone still following this may be interested to participate in discussion #899. Current plan is to include custom headers in the next release, and a PR adding this has been merged.

**Comment by aggregate1166877** [comment]
Holy shit, what an adventure. Thank you so much! Been using a fork, now I can finally use official again.

**Pull request closed without merging**
